File: packages/openview/plugin-callssms/resources/ios/Sources/CallsSmsFunctions.swift
import Foundation
import UIKit
import MessageUI
import Contacts
import ContactsUI

// MARK: - CallsSms Function Namespace

/// Bridge functions for the CallsSms NativePHP plugin on iOS.
///
/// Provides native phone dialling, SMS composition (via MFMessageComposeViewController),
/// WhatsApp deep-link integration, and a CNContactPickerViewController-based contact
/// picker that feeds results back to PHP via LaravelBridge events.
enum CallsSmsFunctions {

    // MARK: - CallsSms.OpenPhoneDialer

    /// Open the native Phone app pre-filled with `phone`.
    ///
    /// Parameters:
    ///   - phone    – phone number string (any format accepted by iOS)
    ///   - autoCall – Bool; true uses `tel://` (immediate call), false uses
    ///                `telprompt://` (shows a confirmation dialog – better UX).
    ///
    /// Both schemes are blocked on devices that cannot place phone calls (e.g. iPad
    /// without cellular); the function returns silently in that case.
    class OpenPhoneDialer: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let phone = parameters["phone"] as? String, !phone.isEmpty else {
                return BridgeResponse.error(message: "phone parameter is required")
            }

            let autoCall = parameters["autoCall"] as? Bool ?? false
            // Percent-encode so any special characters don't break the URL
            let encoded  = phone.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed) ?? phone
            let scheme   = autoCall ? "tel" : "telprompt"

            DispatchQueue.main.async {
                guard let url = URL(string: "\(scheme)://\(encoded)") else { return }

                if UIApplication.shared.canOpenURL(url) {
                    UIApplication.shared.open(url, options: [:], completionHandler: nil)
                } else if autoCall,
                          let fallback = URL(string: "telprompt://\(encoded)"),
                          UIApplication.shared.canOpenURL(fallback) {
                    // telprompt:// sometimes succeeds where tel:// doesn't
                    UIApplication.shared.open(fallback, options: [:], completionHandler: nil)
                }
            }

            return BridgeResponse.success(data: ["success": true])
        }
    }

    // MARK: - CallsSms.OpenSms

    /// Open the native Messages app with `phone` pre-filled as recipient and
    /// `message` as the body text.
    ///
    /// Uses MFMessageComposeViewController (full compose sheet) when available,
    /// otherwise falls back to the `sms:` URL scheme.
    ///
    /// Parameters:
    ///   - phone   – recipient phone number
    ///   - message – (optional) pre-filled message body
    class OpenSms: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let phone = parameters["phone"] as? String, !phone.isEmpty else {
                return BridgeResponse.error(message: "phone parameter is required")
            }

            let message = parameters["message"] as? String ?? ""

            DispatchQueue.main.async {
                if MFMessageComposeViewController.canSendText() {
                    CallsSmsMessageComposer.shared.present(recipient: phone, body: message)
                } else {
                    // Fallback: sms: URL scheme (body may not pre-fill on all iOS versions)
                    let encodedBody = message.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
                    let raw = encodedBody.isEmpty ? "sms:\(phone)" : "sms:\(phone)&body=\(encodedBody)"
                    if let url = URL(string: raw) {
                        UIApplication.shared.open(url, options: [:], completionHandler: nil)
                    }
                }
            }

            return BridgeResponse.success(data: ["success": true])
        }
    }

    // MARK: - CallsSms.OpenWhatsapp

    /// Open WhatsApp in message or call mode for the given `phone`.
    ///
    /// Parameters:
    ///   - phone   – phone in international format ("+15551234567")
    ///   - message – pre-filled text; only used when mode = 'message'
    ///   - mode    – 'message' (default) | 'call'
    ///
    /// Call mode: tries the `whatsapp://call?phone=…` deep link (unofficial but
    /// widely supported as of WhatsApp ≥ 2.x).  Falls back to opening the chat.
    ///
    /// When WhatsApp is not installed the universal `https://wa.me/…` link opens
    /// in the browser and prompts the user to install WhatsApp.
    class OpenWhatsapp: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let phone = (parameters["phone"] as? String)?.trimmingCharacters(in: .whitespaces),
                  !phone.isEmpty else {
                return BridgeResponse.error(message: "phone parameter is required")
            }

            let message = parameters["message"] as? String ?? ""
            let mode    = parameters["mode"]    as? String ?? "message"

            // Normalise: keep leading "+" and strip non-digit characters
            let cleanPhone: String = {
                if phone.hasPrefix("+") {
                    return "+" + phone.dropFirst().filter { $0.isNumber }
                }
                return phone.filter { $0.isNumber }
            }()

            // wa.me requires digits only (no "+")
            let waPhone = cleanPhone.hasPrefix("+") ? String(cleanPhone.dropFirst()) : cleanPhone

            DispatchQueue.main.async {
                let isInstalled = UIApplication.shared.canOpenURL(URL(string: "whatsapp://")!)

                if mode == "call" && isInstalled {
                    // Try the WhatsApp call deep-link (unofficial; works on most current versions)
                    if let callURL = URL(string: "whatsapp://call?phone=\(cleanPhone)"),
                       UIApplication.shared.canOpenURL(callURL) {
                        UIApplication.shared.open(callURL, options: [:], completionHandler: nil)
                        return
                    }
                }

                // Build message / chat URL
                let encodedMsg = message.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
                let urlString: String
                if isInstalled {
                    urlString = encodedMsg.isEmpty
                        ? "whatsapp://send?phone=\(cleanPhone)"
                        : "whatsapp://send?phone=\(cleanPhone)&text=\(encodedMsg)"
                } else {
                    urlString = encodedMsg.isEmpty
                        ? "https://wa.me/\(waPhone)"
                        : "https://wa.me/\(waPhone)?text=\(encodedMsg)"
                }

                if let url = URL(string: urlString) {
                    UIApplication.shared.open(url, options: [:], completionHandler: nil)
                }
            }

            return BridgeResponse.success(data: ["success": true])
        }
    }

    // MARK: - CallsSms.PickContact

    /// Present the system contact picker and dispatch a ContactSelected event
    /// with the chosen name + phone back to PHP.
    ///
    /// Parameters:
    ///   - source – 'device' (default) | 'whatsapp'
    ///              iOS does not expose a WhatsApp-specific contact picker API.
    ///              When source = 'whatsapp' the system contact picker is shown;
    ///              the selected number is then usable with openWhatsapp().
    class PickContact: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let source = parameters["source"] as? String ?? "device"

            DispatchQueue.main.async {
                CallsSmsContactPicker.shared.present(source: source)
            }

            return BridgeResponse.success(data: ["success": true, "message": "Contact picker launched"])
        }
    }

    // MARK: - CallsSms.RequestPermissions

    /// Request all permissions required by this plugin at startup.
    ///
    /// On iOS:
    ///  - Contacts: requests CNContactStore authorization.
    ///    On denial a non-cancellable alert lets the user open Settings or exit.
    ///  - Phone / SMS:  no explicit permission needed (system handles it inline).
    class RequestPermissions: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            DispatchQueue.main.async {
                let store = CNContactStore()
                let status = CNContactStore.authorizationStatus(for: .contacts)

                switch status {
                case .authorized, .limited:
                    // Already granted – nothing to do
                    break
                case .notDetermined:
                    store.requestAccess(for: .contacts) { granted, _ in
                        if !granted {
                            DispatchQueue.main.async {
                                CallsSmsFunctions.showPermissionDeniedAlert()
                            }
                        }
                    }
                case .denied, .restricted:
                    CallsSmsFunctions.showPermissionDeniedAlert()
                @unknown default:
                    break
                }
            }
            return BridgeResponse.success(data: ["success": true, "message": "Permission request initiated"])
        }
    }

    // MARK: - Shared helpers

    /// Present a non-cancellable alert telling the user to open Settings or exit.
    static func showPermissionDeniedAlert() {
        guard let root = UIApplication.shared
            .connectedScenes
            .filter({ $0.activationState == .foregroundActive })
            .compactMap({ $0 as? UIWindowScene })
            .first?.windows.first(where: { $0.isKeyWindow })?.
            rootViewController else { return }

        let alert = UIAlertController(
            title: "Permissions Required",
            message: "This app needs Contacts access to let you pick phone numbers.\n\n"
                   + "Please enable it in Settings to continue.",
            preferredStyle: .alert
        )
        alert.addAction(UIAlertAction(title: "Open Settings", style: .default) { _ in
            if let url = URL(string: UIApplication.openSettingsURLString) {
                UIApplication.shared.open(url)
            }
        })
        alert.addAction(UIAlertAction(title: "Exit App", style: .destructive) { _ in
            exit(0)
        })
        // Not cancellable – user must choose
        root.present(alert, animated: true)
    }
}

File: packages/openview/plugin-callssms/resources/android/CallssmsFunctions.kt
package com.openview.plugins.callssms

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeError
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse

/**
 * Bridge functions for the CallsSms NativePHP plugin on Android.
 *
 * Provides native phone dialler, SMS composition, WhatsApp integration,
 * and a contact picker that dispatches the ContactSelected event back to PHP.
 */
object CallsSmsFunctions {

    // ─────────────────────────────────────────────────────────────────────────
    // CallsSms.OpenPhoneDialer
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Open the device's native phone dialler pre-filled with [phone].
     *
     * Parameters:
     *   phone     – phone number string (any format accepted by the dialler)
     *   autoCall  – Boolean; true attempts an immediate call (ACTION_CALL),
     *               false (default) opens the dialler without calling (ACTION_DIAL).
     *
     * If autoCall = true but CALL_PHONE permission has not been granted the
     * coordinator requests the permission; on grant it places the call, on denial
     * it falls back to ACTION_DIAL so the user can still initiate manually.
     */
    class OpenPhoneDialer(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val phone = parameters["phone"] as? String
                ?: return BridgeResponse.error(BridgeError.InvalidParameters("phone parameter is required"))
            val autoCall = parameters["autoCall"] as? Boolean ?: false

            Handler(Looper.getMainLooper()).post {
                try {
                    if (autoCall) {
                        val hasPermission = ContextCompat.checkSelfPermission(
                            activity, Manifest.permission.CALL_PHONE
                        ) == PackageManager.PERMISSION_GRANTED

                        if (hasPermission) {
                            activity.startActivity(
                                Intent(Intent.ACTION_CALL, Uri.parse("tel:$phone"))
                            )
                        } else {
                            // Request permission; coordinator falls back to ACTION_DIAL on denial
                            val coordinator = CallsSmsCoordinator.install(activity)
                            coordinator.requestCallPermissionAndDial(phone)
                        }
                    } else {
                        activity.startActivity(
                            Intent(Intent.ACTION_DIAL, Uri.parse("tel:$phone"))
                        )
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "OpenPhoneDialer error: ${e.message}", e)
                }
            }

            return BridgeResponse.success(mapOf("success" to true))
        }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // CallsSms.OpenSms
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Open the device's native SMS application pre-filled with [phone] and
     * an optional [message] body.  No runtime permission is needed.
     *
     * Parameters:
     *   phone   – recipient phone number
     *   message – (optional) pre-filled SMS body
     */
    class OpenSms(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val phone = parameters["phone"] as? String
                ?: return BridgeResponse.error(BridgeError.InvalidParameters("phone parameter is required"))
            val message = parameters["message"] as? String ?: ""

            Handler(Looper.getMainLooper()).post {
                try {
                    val intent = Intent(Intent.ACTION_SENDTO, Uri.parse("smsto:$phone")).apply {
                        putExtra("sms_body", message)
                    }
                    activity.startActivity(intent)
                } catch (e: Exception) {
                    Log.e(TAG, "OpenSms error: ${e.message}", e)
                    // Fallback: try plain sms: URI
                    try {
                        val fallback = Intent(Intent.ACTION_VIEW, Uri.parse("sms:$phone")).apply {
                            putExtra("sms_body", message)
                        }
                        activity.startActivity(fallback)
                    } catch (fe: Exception) {
                        Log.e(TAG, "OpenSms fallback error: ${fe.message}", fe)
                    }
                }
            }

            return BridgeResponse.success(mapOf("success" to true))
        }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // CallsSms.OpenWhatsapp
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Open WhatsApp for the given [phone] in either 'message' or 'call' mode.
     *
     * Parameters:
     *   phone   – phone number; international format recommended ("+15551234567")
     *   message – pre-filled text (only used when mode = 'message')
     *   mode    – 'message' (default) | 'call'
     *
     * WhatsApp does not expose a documented deep-link for voice calls on Android.
     * In 'call' mode we open WhatsApp directly to the contact profile/conversation
     * so the user can tap the call button with a single tap.
     */
    class OpenWhatsapp(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val phone = (parameters["phone"] as? String)?.trim()
                ?: return BridgeResponse.error(BridgeError.InvalidParameters("phone parameter is required"))
            val message = parameters["message"] as? String ?: ""
            val mode    = parameters["mode"] as? String ?: "message"

            // Normalise: keep leading '+' and strip everything else that isn't a digit
            val cleanPhone = if (phone.startsWith("+")) {
                "+" + phone.drop(1).filter { it.isDigit() }
            } else {
                phone.filter { it.isDigit() }
            }

            Handler(Looper.getMainLooper()).post {
                try {
                    // Attempt to launch WhatsApp directly.
                    // Using getPackageInfo() to detect WhatsApp fails silently on Android 11+
                    // due to package visibility restrictions, so we try the intent and catch
                    // ActivityNotFoundException instead.
                    when (mode) {
                        "call" -> openWhatsappCallIntent(cleanPhone)
                        else   -> openWhatsappMessageIntent(cleanPhone, message)
                    }
                } catch (e: android.content.ActivityNotFoundException) {
                    Log.d(TAG, "WhatsApp not installed – falling back to wa.me browser")
                    try {
                        val encoded = java.net.URLEncoder.encode(message, "UTF-8")
                        val stripped = cleanPhone.trimStart('+')
                        val url = if (message.isEmpty()) "https://wa.me/$stripped"
                                  else "https://wa.me/$stripped?text=$encoded"
                        activity.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
                    } catch (fe: Exception) {
                        Log.e(TAG, "WhatsApp browser fallback error: ${fe.message}", fe)
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "OpenWhatsapp error: ${e.message}", e)
                }
            }

            return BridgeResponse.success(mapOf("success" to true))
        }

        private fun openWhatsappMessageIntent(phone: String, message: String) {
            val encoded = java.net.URLEncoder.encode(message, "UTF-8")
            val uri = Uri.parse("whatsapp://send?phone=$phone&text=$encoded")
            activity.startActivity(
                Intent(Intent.ACTION_VIEW, uri).setPackage("com.whatsapp")
            )
        }

        /**
         * WhatsApp does not publish an official call deep-link.
         * We open the WhatsApp conversation so the user needs only one tap to call.
         */
        private fun openWhatsappCallIntent(phone: String) {
            val uri = Uri.parse("whatsapp://send?phone=$phone")
            activity.startActivity(
                Intent(Intent.ACTION_VIEW, uri).setPackage("com.whatsapp")
            )
        }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // CallsSms.PickContact
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Open the contact picker and dispatch a ContactSelected event once the user
     * selects a contact.
     *
     * Parameters:
     *   source – 'device' (default) uses the system contact picker;
     *            'whatsapp' uses WhatsApp's built-in contact list (or falls back
     *            to the system picker when WhatsApp is not installed).
     *
     * The result is returned asynchronously via the
     * Openview\Callssms\Events\ContactSelected event dispatched to PHP.
     */
    class PickContact(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val source = parameters["source"] as? String ?: "device"

            Handler(Looper.getMainLooper()).post {
                try {
                    val coordinator = CallsSmsCoordinator.install(activity)
                    coordinator.launchContactPicker(source)
                } catch (e: Exception) {
                    Log.e(TAG, "PickContact error: ${e.message}", e)
                }
            }

            return BridgeResponse.success(mapOf("success" to true, "message" to "Contact picker launched"))
        }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // CallsSms.RequestPermissions
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Request all permissions required by this plugin at once.
     *
     * Permissions requested: CALL_PHONE, READ_CONTACTS.
     *
     * If the user denies any permission a non-cancellable dialog is shown:
     *   • "Grant" re-opens the system permission prompt.
     *   • "Exit"  calls activity.finishAffinity() to close the app.
     *
     * Call this on app startup to ensure all features are available.
     */
    class RequestPermissions(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            Handler(Looper.getMainLooper()).post {
                try {
                    val coordinator = CallsSmsCoordinator.install(activity)
                    coordinator.requestAllPermissions()
                } catch (e: Exception) {
                    Log.e(TAG, "RequestPermissions error: ${e.message}", e)
                }
            }
            return BridgeResponse.success(mapOf("success" to true, "message" to "Permission request initiated"))
        }
    }

    private const val TAG = "CallsSmsFunctions"

    // Legacy stub – kept so any old bridge_functions declarations still resolve
    @Suppress("unused")
    class GetStatus(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            return BridgeResponse.success(mapOf("status" to "ready", "version" to "1.0.0"))
        }
    }
}
